src/macro/index.js

```diff
@@ -0,0 +1,91 @@
+// @flow
+import { createMacro, MacroError } from 'babel-plugin-macros'
+import minify from 'babel-plugin-styled-components/lib/visitors/minify'
+import displayNameAndId from 'babel-plugin-styled-components/lib/visitors/displayNameAndId'
+import templateLiteral from 'babel-plugin-styled-components/lib/visitors/templateLiterals/index'
+import pureAnnotation from 'babel-plugin-styled-components/lib/visitors/pure'
+
+const taggedTemplateImports = ['css', 'keyframes', 'injectGlobal', 'default']
+
+const allowedImports = [
+  'css',
+  'keyframes',
+  'injectGlobal',
+  'isStyledComponent',
+  'consolidateStreamedStyles',
+  'ThemeProvider',
+  'withTheme',
+  'ServerStyleSheet',
+  'StyleSheetManager',
+  '__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS',
+  'default',
+]
+
+function styledComponentsMacro({
+  references,
+  state,
+  babel: { types: t },
+  config = {},
+}) {
+  // create a node for styled-components's imports
+  const program = state.file.path
+  const imports = t.importDeclaration([], t.stringLiteral('styled-components'))
+  // and add it to top of the document
+  program.node.body.unshift(imports)
+
+  // references looks like this
+  // { default: [path, path], css: [path], ... }
+  Object.keys(references).forEach(refName => {
+    if (!allowedImports.includes(refName)) {
+      throw new MacroError(
+        `Invalid import: ${refName}. You can only import ${allowedImports.join(
+          ', '
+        )} from 'styled-components/macro'.`
+      )
+    }
+
+    // add imports
+    let id
+    if (refName === 'default') {
+      id = program.scope.generateUidIdentifier('styled')
+      imports.specifiers.push(t.importDefaultSpecifier(id))
+    } else {
+      id = program.scope.generateUidIdentifier(refName)
+      imports.specifiers.push(t.importSpecifier(id, t.identifier(refName)))
+    }
+
+    references[refName].forEach(referencePath => {
+      // transform the variable imported for the macro
+      // into the new id generated
+      // eslint-disable-next-line no-param-reassign
+      referencePath.node.name = id.name
+
+      if (taggedTemplateImports.includes(refName)) {
+        // find the TaggedTemplateExpression
+        const templatePath = referencePath.findParent(path =>
+          path.isTaggedTemplateExpression()
+        )
+
+        if (!templatePath) {
+          throw new MacroError(
+            `Invalid usage : ${refName}. ${refName} needs to be used with a template literal, like this : ${refName}\`background: red;\`.`
+          )
+        } else {
+          // merge config into the state
+          const stateWithOpts = { ...state, opts: config }
+          // run babel-plugin-styled-components appropriate visitors
+          minify(t)(templatePath, stateWithOpts)
```

Comment:
Is there a way to reuse the setup from the babel plugin wholesale instead of having to recompose all of these? I can't see us maintaining this properly if it isn't more automatic. It's also a bit more complicated as of 1.8 since we're handling call expressions too

Reply:
That sounds like a good idea 😀